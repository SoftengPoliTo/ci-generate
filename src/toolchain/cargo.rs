use std::collections::HashMap;
use std::path::{Path, PathBuf};

use anyhow::Result;
use minijinja::value::Value;

use crate::{builtin_templates, compute_template, define_name_and_license, BuildTemplate};

static CARGO_TEMPLATES: &[(&str, &str)] = &builtin_templates!["cargo" =>
    ("md.README", "README.md"),
    ("ci.gitlab", ".gitlab-ci.yml"),
    ("ci.github.compact", "github-compact.yml"),
    ("ci.github.ubuntu", "github-ubuntu.yml"),
    ("ci.github.macos", "github-macos.yml"),
    ("ci.github.windows", "github-windows.yml"),
    ("ci.github.deploy", "github-deploy.yml")
];

/// A cargo project data.
pub struct Cargo;

impl Cargo {
    /// Creates a new CI for a cargo project.
    pub fn create_ci(project_path: &Path, license: &str) -> Result<()> {
        let (project_name, license) = define_name_and_license(project_path, license)?;
        let template = Cargo.build(project_path, project_name, license.id());
        compute_template(template, license)
    }

    fn project_structure(
        project_path: &Path,
        name: &str,
    ) -> (HashMap<PathBuf, &'static str>, Vec<PathBuf>) {
        let root = project_path.to_path_buf();
        let github = project_path.join(".github/workflows");

        let mut template_files = HashMap::new();

        // README
        template_files.insert(root.join("README.md"), "md.README");

        // Continuous Integration
        template_files.insert(root.join(".gitlab-ci.yml"), "ci.gitlab");
        template_files.insert(github.join(format!("{}.yml", name)), "ci.github.compact");
        template_files.insert(
            github.join(format!("{}-ubuntu.yml", name)),
            "ci.github.ubuntu",
        );
        template_files.insert(
            github.join(format!("{}-macos.yml", name)),
            "ci.github.macos",
        );
        template_files.insert(
            github.join(format!("{}-windows.yml", name)),
            "ci.github.windows",
        );
        template_files.insert(github.join("deploy.yml"), "ci.github.deploy");

        (template_files, vec![root, github])
    }
}

impl BuildTemplate for Cargo {
    fn define(
        &self,
        project_path: &Path,
        project_name: &str,
        license: &str,
    ) -> (
        HashMap<PathBuf, &'static str>,
        Vec<PathBuf>,
        HashMap<&'static str, Value>,
    ) {
        let mut context = HashMap::new();

        context.insert("name", Value::from_serializable(&project_name));
        context.insert("license_id", Value::from_serializable(&license));

        let (files, dirs) = Cargo::project_structure(project_path, project_name);

        (files, dirs, context)
    }

    fn get_templates() -> &'static [(&'static str, &'static str)] {
        CARGO_TEMPLATES
    }
}
