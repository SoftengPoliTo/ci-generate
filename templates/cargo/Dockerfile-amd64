FROM amd64/alpine
ADD {{ name }} /
ENTRYPOINT ["/{{ name }}"]
