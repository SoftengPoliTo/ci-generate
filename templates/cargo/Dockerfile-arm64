FROM arm64v8/alpine
ADD {{ name }} /
ENTRYPOINT ["/{{ name }}"]
