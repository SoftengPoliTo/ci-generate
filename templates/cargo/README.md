# {{ name }}

[![Actions Status](https://github.com/<your-account>/{{ name }}/workflows/{{ name }}/badge.svg)](https://github.com/<your-account>/{{ name }}/actions)
[![CodeCov](https://codecov.io/gh/<your-account>/{{ name }}/branch/master/graph/badge.svg)](https://codecov.io/gh/<your-account>/{{ name }})
[![LICENSE](https://img.shields.io/badge/license-{{ license_id }}-blue.svg)](LICENSE)

{{ name }}'s description

# Acknowledgements
