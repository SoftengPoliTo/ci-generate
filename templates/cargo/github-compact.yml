name: {{ name }}

on:
  push:
    branches:
      - {{ branch }}
  pull_request:
    branches:
      - {{ branch }}

jobs:
  clippy-rustfmt:

    strategy:
      matrix:
        platform: [ubuntu-latest, macos-latest, windows-latest]

    runs-on: {{ '${{ matrix.platform }}' }}

    steps:
    - uses: actions/checkout@v3

    - name: Install Rust stable
      uses: dtolnay/rust-toolchain@stable
      with:
        profile: minimal
        toolchain: stable
        override: true
        components: clippy, rustfmt

    - name: Run rustfmt
      run:
        cargo fmt --all -- --check --verbose

    - name: Run cargo clippy
      uses: giraffate/clippy-action@v1
      with:
        github_token: {{ '${{ secrets.GITHUB_TOKEN }}' }}
        clippy_flags: --all-targets -- -D warnings
        reporter: github-pr-review

  build-test:

    strategy:
      matrix:
        platform: [ubuntu-latest, macos-latest, windows-latest]

    runs-on: {{ '${{ matrix.platform }}' }}

    steps:
    - uses: actions/checkout@v3

    - name: Install Rust stable
      uses: dtolnay/rust-toolchain@stable
      with:
        profile: minimal
        toolchain: stable
        override: true

    - name: Build
      run: cargo build --verbose --all-targets

    - name: Run tests
      run: cargo test --verbose

    - name: Generate docs
      run: cargo doc --no-deps

