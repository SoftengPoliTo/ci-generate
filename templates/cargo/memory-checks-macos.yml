name: memory-checks-macos

on:
  push:
    branches:
      - {{ branch }}
  pull_request:
    branches:
      - {{ branch }}

jobs:
  undefined-behaviour-fuzzy-dynamic-analysis:

    runs-on: macos-latest

    steps:
    - uses: actions/checkout@v3

    - name: Cache produced data
      id: cache-data
      uses: actions/cache@v3
      with:
        path: |
          ~/.cargo/bin/
          ~/.cargo/registry/index/
          ~/.cargo/registry/cache/
          ~/.cargo/git/db/
          target/
        key: {{ '${{ runner.os }}' }}-cargo-ci-{{ "${{ hashFiles('**/Cargo.toml') }}" }}

    - name: Install Rust nightly and miri
      uses: dtolnay/rust-toolchain@stable
      with:
        toolchain: nightly
        components: miri, rust-src
        override: true

    # FIXME Use binaries
    - name: Install cargo-fuzz
      if: steps.cache-data.outputs.cache-hit != 'true'
      run: |
        cargo install cargo-fuzz

    - name: Run miri
      env:
        # -Zrandomize-layout makes sure not to rely on the layout of anything
        # that might change
        RUSTFLAGS: -Zrandomize-layout
        # -Zmiri-symbolic-alignment-check makes the alignment check more strict.
        MIRIFLAGS: >
          -Zmiri-symbolic-alignment-check
      run: cargo miri test

    # FIXME Create a template with a dummy series of fuzzy tests
    - name: Init cargo-fuzz
      run: cargo fuzz init

    - name: Run cargo-fuzz
      run: cargo fuzz build

    # Usage of the `help` command as base command, please replace it
    # with the effective command that AddressSanitizer has to analyze
    - name: Run AddressSanitizer
      env:
        RUSTFLAGS: -Zsanitizer=address -Copt-level=3
        RUSTDOCFLAGS: -Zsanitizer=address
      run: cargo test -Zbuild-std --target x86_64-apple-darwin
      # Use `cargo run` for the analysis of a binary.
      # Usage of the `help` command as base command, please replace it
      # with the effective command that AddressSanitizer has to analyze
      # run: cargo run -Zbuild-std --target x86_64-apple-darwin -- --help
