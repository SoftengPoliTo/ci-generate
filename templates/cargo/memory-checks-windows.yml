name: memory-checks-windows

on:
  push:
    branches:
      - {{ branch }}
  pull_request:
    branches:
      - {{ branch }}

jobs:
  undefined-behaviour-dynamic-analysis:

    runs-on: windows-latest

    steps:
    - uses: actions/checkout@v3

    - name: Cache produced data
      id: cache-data
      uses: actions/cache@v3
      with:
        path: |
          ~/.cargo/bin/
          ~/.cargo/registry/index/
          ~/.cargo/registry/cache/
          ~/.cargo/git/db/
          target/
        key: {{ '${{ runner.os }}' }}-cargo-ci-{{ "${{ hashFiles('**/Cargo.toml') }}" }}

    - name: Install Rust nightly and miri
      uses: dtolnay/rust-toolchain@stable
      with:
        toolchain: nightly
        components: miri
        override: true

    - name: Run miri
      env:
        # -Zrandomize-layout makes sure not to rely on the layout of anything
        # that might change
        RUSTFLAGS: -Zrandomize-layout
        # -Zmiri-symbolic-alignment-check makes the alignment check more strict.
        MIRIFLAGS: >
          -Zmiri-symbolic-alignment-check
      run: cargo miri test
