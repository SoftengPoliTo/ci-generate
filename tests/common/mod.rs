use ci_generate::cargo::SKIPPED_FOLDERS;
use std::{fs, path::Path};
use walkdir::WalkDir;

pub fn compare_template(snapshot_path: &Path, template_path: &Path) {
    // https://docs.rs/walkdir/latest/walkdir/struct.IntoIter.html#method.skip_current_dir
    let mut it = WalkDir::new(template_path).into_iter();
    loop {
        let entry = match it.next() {
            None => break,
            Some(entry) => entry.unwrap(),
        };
        let file_name = entry.file_name().to_string_lossy().to_string();
        let skip_entry = SKIPPED_FOLDERS.contains(&file_name.as_str());

        if skip_entry && entry.file_type().is_dir() {
            it.skip_current_dir();
            continue;
        }
        if entry.file_type().is_file() && !SKIPPED_FOLDERS.contains(&file_name.as_str()) {
            compare(snapshot_path, template_path, entry.path());
        }
    }
}

fn compare(snapshot_path: &Path, path: &Path, entry: &Path) {
    let content = fs::read_to_string(entry).unwrap();
    let name = entry.file_name().and_then(|v| v.to_str());
    insta::with_settings!({
        snapshot_path => snapshot_path
        .join(entry.strip_prefix(path).unwrap())
        .parent()
        .unwrap(),
        prepend_module_to_snapshot => false,
    },{
        insta::assert_snapshot!(name, content);
    });
}
