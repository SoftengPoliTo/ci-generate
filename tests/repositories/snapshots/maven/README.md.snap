---
source: tests/common/mod.rs
expression: content
---
# Maven-project

[![Actions Status][actions badge]][actions]
[![CodeCov][codecov badge]][codecov]
[![LICENSE][license badge]][license]

Maven-project's description

# Acknowledgements

<!-- Links -->
[actions]: https://github.com/<your-account>/Maven-project/actions
[codecov]: https://codecov.io/gh/<your-account>/Maven-project
[license]: LICENSES/BSD-1-Clause.txt

<!-- Badges -->
[actions badge]: https://github.com/<your-account>/Maven-project/workflows/Maven-project/badge.svg
[codecov badge]: https://codecov.io/gh/<your-account>/Maven-project/branch/master/graph/badge.svg
[license badge]: https://img.shields.io/badge/license-BSD-1-Clause-blue.svg
